lightning/src/chain/mod.rs

```diff
@@ -262,37 +217,44 @@ pub enum ChannelMonitorUpdateStatus {
-/// without taking any further action such as persisting the current state.
+/// completion. If an update will not succeed, then it must immediately shut down.
 ///
 /// If an implementation maintains multiple instances of a channel's monitor (e.g., by storing
 /// backup copies), then it must ensure that updates are applied across all instances. Otherwise, it
 /// could result in a revoked transaction being broadcast, allowing the counterparty to claim all
 /// funds in the channel. See [`ChannelMonitorUpdateStatus`] for more details about how to handle
 /// multiple instances.
-///
-/// [`PermanentFailure`]: ChannelMonitorUpdateStatus::PermanentFailure
 pub trait Watch<ChannelSigner: WriteableEcdsaChannelSigner> {
 	/// Watches a channel identified by `funding_txo` using `monitor`.
 	///
 	/// Implementations are responsible for watching the chain for the funding transaction along
 	/// with any spends of outputs returned by [`get_outputs_to_watch`]. In practice, this means
 	/// calling [`block_connected`] and [`block_disconnected`] on the monitor.
 	///
-	/// Note: this interface MUST error with [`ChannelMonitorUpdateStatus::PermanentFailure`] if
-	/// the given `funding_txo` has previously been registered via `watch_channel`.
+	/// A return of `Err(())` indicates that the channel should not open and the channel
+	/// immediately force-closed without broadcasting the funding transaction.
+	///
+	/// If the given `funding_txo` has previously been registered via `watch_channel`, `Err(())`
+	/// must be returned.
 	///
 	/// [`get_outputs_to_watch`]: channelmonitor::ChannelMonitor::get_outputs_to_watch
 	/// [`block_connected`]: channelmonitor::ChannelMonitor::block_connected
 	/// [`block_disconnected`]: channelmonitor::ChannelMonitor::block_disconnected
-	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<ChannelSigner>) -> ChannelMonitorUpdateStatus;
+	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<ChannelSigner>) -> Result<ChannelMonitorUpdateStatus, ()>;
 
 	/// Updates a channel identified by `funding_txo` by applying `update` to its monitor.
 	///
-	/// Implementations must call [`update_monitor`] with the given update. See
-	/// [`ChannelMonitorUpdateStatus`] for invariants around returning an error.
+	/// Implementations must call [`ChannelMonitor::update_monitor`] with the given update. This
+	/// may fail (returning an `Err(())`), in which case this should return
+	/// [`ChannelMonitorUpdateStatus::InProgress`] (and the update should never complete). This
+	/// generally implies the channel has been closed (either by the funding outpoint being spent
+	/// on-chain or the [`ChannelMonitor`] having decided to do so and broadcasted a transaction),
+	/// and the [`ChannelManager`] state will be updated once it sees the funding spend on-chain.
+	///
+	/// If persistence fails, this should return [`ChannelMonitorUpdateStatus::InProgress`] and
```

Comment:
>  I also wonder if we need to rename InProgress to break backwards compat of the type to force users to check the docs as they now have much more responsibilities?

+1. I had mentioned changing this to `Pending` in a previous discussion. I also think the meaning is slightly different, though this might just be my context for the terms. `InProgress` sounds like this is something actively being worked on, while `Pending` sounds like something that is blocked from progress on another event, which I think is closer to your meaning now. Under normal operation, it's simply "pending" a successful write, but when things go wrong it's "pending" a more sophisticated response.

Reply:
I mostly wrote that because if you _do_ need to search for a new name to force re-examination of the API, it's a candidate. `Pending` and `InProgress` are closer to synonymous than not.